preemptive_requests.py

```diff
@@ -35,22 +121,120 @@ def setup_requests(self):
+                # insert a fake customized QTableWidgetItem to allow sorting
+                # based on Transmission - column position 1
+                trans_item = CustomTableWidgetItem(
+                    widget_type=QtWidgets.QLabel,
+                    widget_name='transmission_label')
+                trans_item.setSizeHint(widget.size())
+                reqs_table.setItem(row_position, 1, trans_item)
+
+                count += 1
+        reqs_table.resizeRowsToContents()
+        self.update_filters()
         print(f'Added {count} preemptive requests')
 
+    def sort_rate_items(self, value):
+        """
+        Sort the items in the table based on the Rate values from the embedded
+        widget. sortItems will call CustomTableWidgetItem's __ld__ method where
+        the values from the rate label (CustomTabWidgetItem column 0) will be
+        compared and sorted.
+        """
+        column = 0
+        if value is True:
+            self.ui.reqs_table_widget.sortItems(column,
+                                                QtCore.Qt.DescendingOrder)
+        else:
+            self.ui.reqs_table_widget.sortItems(column,
+                                                QtCore.Qt.AscendingOrder)
+
+    def sort_transmission_items(self, value):
+        """
+        Sort the items in the table based on the Transmission values from the
+        embedded widget. sortItems will call CustomTableWidgetItem's __ld__
+        method, where the values from the transmission label
+        (CustomTabWidgetItem column 1) will be compared and sorted.
+        """
+        column = 1
+        if value is True:
+            self.ui.reqs_table_widget.sortItems(column,
+                                                QtCore.Qt.DescendingOrder)
+        else:
+            self.ui.reqs_table_widget.sortItems(column,
+                                                QtCore.Qt.AscendingOrder)
+
+    def enable_bits(self, toggle):
+        """
+        Enable the bits when the combo box is checked.
+        """
+        if toggle is True:
+            for key, item in self._bits.items():
+                cb = self.findChild(QtWidgets.QCheckBox, f"filter_cb_{key}")
+                cb.setEnabled(True)
+        else:
+            for key, item in self._bits.items():
+                cb = self.findChild(QtWidgets.QCheckBox, f"filter_cb_{key}")
+                cb.setEnabled(False)
+
+    def calc_bitmask(self):
+        """
+        Calculate a decimal number based on the checked/unchecked bits
+        from the filter (Photon Energy Range).
+        These bits are represented by check boxes where checked = True and
+        unchecked = False.
+        """
+        for key, item in self._bits.items():
+            cb = self.findChild(QtWidgets.QCheckBox, f"filter_cb_{key}")
+            self._bits[key] = cb.isChecked()
+
+        bit_map = list(map(int, [item for key, item in self._bits.items()]))
```

Comment:
Are we using `pmps-ui` - with python < 3.6 somewhere? - if so I'm in trouble here and need to change this because I think the dictionaries are not ordered in previous python versions.

Reply:
pydm is not using versions < 3.6 so i think I am safe here, unless we use it with Python 2 ?

Reply:
We're running live with python=3.6 in dev_conda and python=3.8 in the tagged version